packages/langium/src/documents/document-builder.ts

```diff
@@ -44,65 +40,78 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
     protected readonly documentValidator: DocumentValidator;
     protected readonly langiumDocuments: LangiumDocuments;
     protected readonly indexManager: IndexManager;
+    protected readonly linker: Linker;
 
     constructor(services: LangiumServices) {
         this.connection = services.lsp.Connection;
         this.parser = services.parser.LangiumParser;
+        this.linker = services.references.Linker;
         this.scopeComputation = services.references.ScopeComputation;
         this.documentValidator = services.validation.DocumentValidator;
         this.langiumDocuments = services.documents.LangiumDocuments;
         this.indexManager = services.index.IndexManager;
     }
 
-    build(document: LangiumDocument): BuildResult {
-        const parseResult = this.parser.parse(document);
-        document.parseResult = parseResult;
-        this.process(document);
-        let diagnostics: Diagnostic[] | undefined = this.connection ? this.validate(document) : undefined;
-        const validator = this.documentValidator;
+    async build(document: LangiumDocument, cancelToken = CancellationToken.None): Promise<BuildResult> {
+        await this.buildDocument(document, cancelToken);
         return {
             parseResult: document.parseResult,
```

Comment:
The `parseResult` is not built in this function, but is already contained in the given document. We should remove it from the BuildResult.